### ISSUE TITLE
For Printing

### ISSUE THREAD

**Issue by shiffman** [CodingTrain/CommunityClouds#50]
Thank you @TurqDevDesign for the amazing SVG generator in #46!

I have some clarification about the plan for Processing day from @tchoi8.

We'll want all the shapes to fit  4 x 3" cards so ideally we'll create some sort of generator that makes a PDF with 6 random clouds on each page.

The clouds should be black and white and not include the name text as these will be handwritten.

![image 3](https://user-images.githubusercontent.com/191758/31091793-1894d6a8-a77a-11e7-9e0d-8eb06ffbff08.png)

Would generating images actually end up being easier?  Does it make more sense to create a separate app, not part of the github pages site, to generate the designs from `generators.js`?

**Comment by meiamsome** [comment]
I think the best plan for this would be:

- [ ] Mark up clouds that are purely white-fill black/gray stroke.
- [ ] Edit the SVG generation to allow generating without background and text (Should be simple).
- [ ] Make a way to request n random SVGs of a specific dimension.
- [ ] Write a small program to find the SVG's bounding boxes (Which would be the cloud's bounding boxes) and scale them up to an appropriate size. (Or find a useable one)
- [ ] Position 6 SVG on one page somehow.

If anyone wants to take on any part of this, please drop a comment in here!

**Comment by shiffman** [comment]
Great plan, can we set a deadline for this next Monday, Oct 9th? Maybe we should add something to the README as a deadline for getting a cloud in to have a chance of it being included in the printed nametags (though this repo can remain open for pull requests indefinitely). I can mention on Friday's livestream also.

**Comment by TurqDevDesign** [comment]
You're welcome @shiffman! A couple questions, do you want the clouds to be purely black and white? Or would you be all right with greyscale? Second, will all of the qualifying clouds have to have an outline, and be on a white background? Lastly, if the previous question is correct, does that mean that we'll have to pick out all of the clouds that fit this criteria?

Also, @meiamsome let me do a little quick research, this sounds like a fun challenge, but I need to make sure it's possible in the manner I'm thinking 😜

**Comment by shiffman** [comment]
Maybe @tchoi8 can way in on the black & white vs grayscale question. It just depending on how he's planning on doing the printing. 

I believe the clouds should be an outline on a white background yes, the names will be handwritten in and there has to be enough space left on the card!

**Comment by tchoi8** [comment]
Hi all, 

Thanks for great contributions. We are so excited.

- do you want the clouds to be purely black and white? Or would you be all right with greyscale?
: B&W will work and Greyscale are fine. 

-  Second, will all of the qualifying clouds have to have an outline, and be on a white background? 
: Yes, it's best if the clouds have an outline, as that will give us most flexibility for printing. 

- Lastly, if the previous question is correct, does that mean that we'll have to pick out all of the clouds that fit this criteria?
: I think this is a curatorial decision.

**Comment by shiffman** [comment]
I'm happy to take responsibility for the curation. In an ideal world, I'd like everyone's cloud to be included in at least one name tag, but if it does not meet the technical criteria, we'll have to leave it out. What if we print the github username that created the cloud in a small font in the corner of each name tag?

**Comment by tchoi8** [comment]
Hi @meiamsome the [generator](https://codingtrain.github.io/CommunityClouds/print/) is awesome.  

Would you be interested in adding an option to download/ print generated designs into a single image/ PDF with six 4 x 3" cards?

We need to make about 250 unique designs and print them on this [label template](https://www.uline.com/Product/Detail/S-10446/Badges/Name-Badge-Holders-Additional-Inserts-4-x-3).  
Or, if you have a suggestion on how to save the generated designs from the blob into individual svg files, please let us know!

thanks

**Comment by meiamsome** [comment]
@tchoi8 So, currently they're output in the 2x3 grid of 4x3" cards. If you know some accurate measurements of the borders for the printing I can burn these into the SVG so that they are output correctly as they should be.

Currently you can download the blob of six of them for a page just as an SVG - but I can add print styling so you can use most browsers to print to PDF if we can get the exact spacing around them

**Comment by meiamsome** [comment]
Ah, I think I just found the spacing on their website!

**Comment by meiamsome** [comment]
@tchoi8 Okay, would you like to check if that's okay? 
You may need chrome/firefox for the best operation of the following:
Click the add one of each button and then go to the print dialogue - set all margins to 0 and the paper size to 8.5x11" and then print to PDF and it should come out in the exact printable size.

Let me know if that works for you.

**Comment by tchoi8** [comment]
Hi @meiamsome thanks so much, and sorry for slow reply. we got to try printing the design with name tag paper. 

![nametag_template](https://user-images.githubusercontent.com/683107/31736604-40c71406-b413-11e7-9f36-3a3518896448.jpg)

As you can see there's a slight offset in the placement. I think it's due to the paper being actually different from the layout they make available online. 

We marked the dimensions of cut line as a reference. Would you be able to take a look? 
Also, would you point me to the part of the code that's adjusting layout? It may come handy to know about it as we make small adjustments before the printing. 

Thanks so much. 

We printed a few out as a test by manually adjusting the layout. I think they will be fabulous! 
 
![unnamed](https://user-images.githubusercontent.com/683107/31736751-94149174-b413-11e7-882f-3587d9ae5919.jpg)

**Comment by meiamsome** [comment]
@tchoi8 Yes, I got some things wrong when adapting the dimensions (Some miscalculations here and there...) It should be correct now. I'm not sure if your browser is printing without margins correctly judging by your picture, here is  a pdf of the page as rendered through Firefox (Which allows you to set the paper dimensions to be custom... The units are all in mm for me so I did 850x1100mm which should scale correctly and set all margins to zero. If this is correct and you cannot get the dimensions correct on your browser (It was a pain on all the ones I've tried) I would be happy to send you just a bunch of PDFs of clouds.

[mozilla.pdf](https://github.com/CodingTrain/CommunityClouds/files/1396037/mozilla.pdf)

Oh, and those name tags look great!

**Comment by tchoi8** [comment]
@meiamsome  We tried and they fit much better except the clouds on the left are too close to the center by about 0.25” and the clouds in the bottom row are hitting the top edge of the cards.
 
Since the clouds are shared so uniquely, their margins may need to be set differently for each type of clouds. I see this [commit](https://github.com/CodingTrain/CommunityClouds/commit/7a7b52c77bc1d45fc78d4a5bc4a599e5d45ecbdb#diff-6fcb1e287ed20ff9ca6993e5068edc05) highlights the numbers that can be tweaked. 

I think we are set for now, I will try printing with more papers on site and report back with results or more questions. Thanks!

**Comment by meiamsome** [comment]
@tchoi8 I can also adjust the internal margins fairly simply if that would be better for making sure they fit

**Comment by tchoi8** [comment]
@meiamsome THANK YOU. We printed 250 clouds and will use them tomorrow for Processing Community Day.  

![img_8847](https://user-images.githubusercontent.com/683107/31831932-258bfacc-b593-11e7-9222-5dbca2fa32f5.JPG)
A small edit. The print on the bottom left show cloud 5 and 6 a bit close to the cut line. I edited the print.js code to output height to be 420 and made a condition for cloud 5 and 6 for outputHeight + 520, to fit most clouds on the paper. 

 I think this issue can be closed now. Thanks so much.
